iothub/device/devdoc/API Design.md

```diff
@@ -0,0 +1,302 @@
+{
+    PropertyValue = propertyValue;
+    AckCode = ackCode;
+    AckVersion = ackVersion;
+    AckDescription = ackDescription;
+}
+
+/// <summary>
+/// The unserialized property value.
+/// </summary>
+[JsonProperty("value")]
+public object PropertyValue { get; set; }
+
+/// <summary>
+/// The acknowledgement code, usually an HTTP Status Code e.g. 200, 400.
+/// </summary>
+[JsonProperty("ac")]
+public int AckCode { get; set; }
+
+/// <summary>
+/// The acknowledgement version, as supplied in the property update request.
+/// </summary>
+[JsonProperty("av")]
+public long AckVersion { get; set; }
+
+/// <summary>
+/// The acknowledgement description, an optional, human-readable message about the result of the property update.
+/// </summary>
+[JsonProperty("ad", DefaultValueHandling = DefaultValueHandling.Ignore)]
+public string AckDescription { get; set; }
+```
+</details>
+
+
+### Telemetry
+
+```csharp
+/// <summary>
+/// Sends a single instance of telemetry.
+/// </summary>
+/// <param name="telemetryName">The name of the telemetry to send.</param>
+/// <param name="telemetryValue">The value of the telemetry to send.</param>
+/// <param name="conventionHandler">A convention handler that defines the content encoding and serializer to use for the telemetry message.</param>
+/// <param name="componentName">The component name this telemetry belongs to.</param>
+/// <param name="cancellationToken">A cancellation token.</param>
+/// <remarks>
+/// This will create a single telemetry message and will not combine multiple calls into one message. Use <seealso cref="SendTelemetryAsync(IDictionary{string, dynamic}, string, IConventionHandler, CancellationToken)"/>. Refer to the documentation for <see cref="IConventionHandler"/> if you want to use a custom serializer.
+/// </remarks>
+public Task SendTelemetryAsync(string telemetryName, object telemetryValue, string componentName = default, IConventionHandler conventionHandler = default, CancellationToken cancellationToken = default);
+
+/// <summary>
+/// Sends a collection of telemetry.
+/// </summary>
+/// <param name="telemetryDictionary">A dictionary of dynamic objects </param>
+/// <param name="componentName">The component name this telemetry belongs to.</param>
+/// <param name="conventionHandler">A convention handler that defines the content encoding and serializer to use for the telemetry message.</param>
+/// <param name="cancellationToken">A cancellation token.</param>
+/// /// <remarks>
+/// This will either use the <see cref="DefaultConvention"/> to define the encoding and use the default Json serailzier. Refer to the documentation for <see cref="IConventionHandler"/> if you want to use a custom serializer.
+/// </remarks>
+public Task SendTelemetryAsync(IDictionary<string, object> telemetryDictionary, string componentName = default, IConventionHandler conventionHandler = default, CancellationToken cancellationToken = default);
```

Comment:
So what this would do is to take the IDictinory<string,object> and turn it into some big JSON blob to be sent over, correct?  So like "<"temp",1234>, <"humid",32>" gets turned into {"temp":1234, "humid":32}, correct?

I think we should really think about whether the V1 of API needs this?  It doesn't seem that hard for an application to build up this JSON own their own, even potentially with an Object they own that straight serializes to JSON (so like temp and humid could both be members with [JSON] rules right?).  

So I don't really think this API is needed, it's yet another overload, and especially if there's say need for a CBOR serilaizer at some point we - and not application - end up taking over the responsibility of figuring all this out.

Reply:
Similar though not quite as strong feedback about the `SendTelemetryAsync(string telemetryName, object telemetryValue` overload. 

For dealing with properties, having the SDK have serialization logic adds a bunch of value since there's all that metadata gunk that needs dealing with.  But on telemetry this is just some name:value pair that you can look directly at the DTDL and guess how the underlying body would work.

Reply:
I want to make sure that I understand this correctly - do we want the Message class to only accept the payload bytes and not care about how those bytes were generated?

Reply:
Yes, exactly.  The `Message` class seems pretty straightforward as/is to use, we'd just need to add Component as a top-level component since that is PnP centric.  But otherwise this isn't that hard to serialize at app layer so I think we should leave there.  

C is definitely following this model (we don't have same concept of serializers floating around as common pattern) but on this one I think other languages can keep simple too.